cng/cng.go

```diff
@@ -36,32 +36,35 @@ func len32(s []byte) int {
 
 var algCache sync.Map
 
-type newAlgEntryFn func(h bcrypt.ALG_HANDLE) (interface{}, error)
-
-func loadOrStoreAlg(id string, flags bcrypt.AlgorithmProviderFlags, mode string, fn newAlgEntryFn) (interface{}, error) {
+// loadOrStoreAlg loads an algorithm with the given id, flags, and mode from the cache.
+// If the algorithm is not in the cache, a new one is created and then initialized using fn.
+// The returned algorithm handle should not be closed by the caller.
+func loadOrStoreAlg[T any](id string, flags bcrypt.AlgorithmProviderFlags, mode string, fn func(h bcrypt.ALG_HANDLE) (T, error)) (T, error) {
```

Comment:
Would it be idiomatic in Go to split out the load and store logic into their own small private methods with this function as the wrapper?  Not sure if that could/would introduce a race condition if it wasn't all in the one function.

Reply:
Calling other funcs from in here is fine. I'm not totally sure what split you're suggesting though, and what the purpose would be. FWIW, I find this func easy to read. (Being self-contained helps with that, but of course that is just a factor.)

Maybe what you have in mind is something like:

```
func loadOrStore[V any](m *sync.Map, key any, fn func() (V, error)) (V, error)
```

And then `loadOrStoreAlg` could set up `key` and `fn` and call `loadOrStore`. If we had multiple of these, I think that would make sense. I'm not sure the current case is in itself complicated enough to justify it though.

Reply:
> Would it be idiomatic in Go to split out the load and store logic into their own small private methods with this function as the wrapper?

This function doesn't follow any specific convention, but it is idiomatic. We could implement the load and store operations in different functions, combining both in the `loadOrStoreAlg`, that would also be idiomatic. I haven't done this because we currently don't need to call these operations separately, and I would rather not see callers trying to reimplement the load or store logic, as it is difficult to do it race-free.